db_sql.go

```diff
@@ -123,22 +126,71 @@ func (x *sqlUserStoreDB) ResetPasswordFinish(userId UserId, token string, passwo
 			}
 			return escan
 		}
+				
 		if everify := verifyPasswordResetToken(token, truthToken.String); everify != nil {
 			tx.Rollback()
 			return everify
 		}
+
+		if used := x.verifyPasswordArchived(userId, password); used {
+			tx.Rollback()
+			return ErrInvalidPastPassword
+		}
+
+		if earchive := x.archivePassword(tx, userId); earchive != nil {
+			tx.Rollback()
+			return earchive
+		}
+
 		if eupdate := x.setPasswordInternal(tx, userId, password); eupdate != nil {
 			tx.Rollback()
 			return eupdate
-		} else {
-			return tx.Commit()
 		}
-
+		return tx.Commit()
 	} else {
 		return etx
 	}
 }
 
+func (x *sqlUserStoreDB) archivePassword(tx *sql.Tx, userId UserId) error {
+	if _, err := tx.Exec(`INSERT INTO authpwdarchive (userid, password) SELECT userid, password FROM authuserpwd WHERE userId = $1`, userId); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (x *sqlUserStoreDB) verifyPasswordArchived(userId UserId, password string) bool {
```

Comment:
The word 'verify' doesn't tell us what the function's return value means. It could rather be named something like `hasPasswordBeenUsedBefore`